src/sage/rings/valuation/valuation_space.py

```diff
@@ -0,0 +1,1574 @@
+                sage: v._test_bounds()
+
+            """
+            tester = self._tester(**options)
+
+            X = self.domain().some_elements()
+            for x in tester.some_elements(X):
+                tester.assertGreaterEqual(self.upper_bound(x), self(x))
+                tester.assertLessEqual(self.lower_bound(x), self(x))
+
+        def _test_simplify(self, **options):
+            r"""
+            Check that :meth:`simplify` works correctly.
+
+            TESTS::
+
+                sage: v = ZZ.valuation(3)
+                sage: v._test_simplify()
+
+            """
+            tester = self._tester(**options)
+
+            try:
+                k = self.residue_ring()
+                has_residue_ring = True
+            except NotImplementedError:
+                # over non-fields (and especially polynomial rings over
+                # non-fields) computation of the residue ring is often
+                # difficult and not very interesting
+                from sage.categories.fields import Fields
+                if self.domain() not in Fields():
+                    return
+                raise
+
+            X = self.domain().some_elements()
+            for x in tester.some_elements(X):
+                y = self.simplify(x)
+                tester.assertEqual(self(x), self(y))
+                if self(x) >= 0 and  has_residue_ring:
+                    tester.assertEqual(self.reduce(x), self.reduce(y))
+
+            if self.is_trivial() and not self.is_discrete_valuation():
+                return
+
+            S = self.value_group().some_elements()
+            from itertools import product
+            for x,s in tester.some_elements(product(X, S)):
+                y = self.simplify(x, error=s)
+                if self.domain().is_exact():
+                    tester.assertGreaterEqual(self(x-y), s)
+                elif hasattr(y, 'precision_absolute'):
+                    tester.assertGreaterEqual(self(x-y), min(s, y.precision_absolute()))
+
+        def _test_shift(self, **options):
+            r"""
+            Check that :meth:`shift` works correctly.
+
+            TESTS::
+
+                sage: v = ZZ.valuation(3)
+                sage: v._test_shift()
+
+            """
+            if self.is_trivial() and not self.is_discrete_valuation():
+                return
+
+            try:
+                self.residue_ring()
+            except:
+                # it is not clear what a shift should be in this case
+                return
+
+            tester = self._tester(**options)
+            X = self.domain().some_elements()
+            S = self.value_group().some_elements()
+            from itertools import product
+            for x,s in tester.some_elements(product(X, S)):
+                if self(x) < 0 and ~self.uniformizer() not in self.domain():
+                    # it is not clear what a shift should be in this case
+                    continue
+                self.shift(x, s)
```

Comment:
It might be worth testing something here beyond just shift not raising an error.  For example, in exact domains if you shift up and then down, you should return to where you started.